### PR TITLE
support chatglm3

### PR DESCRIPTION
**Pull request by gameofdimension**
### correctness

we test this pr on tp=1/2/4, all emit reasonable output

> Prompt: 'Hello, my name is', Generated text: ' 👨\u200d💻 and I am a computer program'
> Prompt: 'The president of the United States is', Generated text: ' 45 years old. The vice president of the United States is 4'
> Prompt: 'The capital of France is', Generated text: " 100 miles from the center of the Earth. If the Earth'"
> Prompt: 'The future of AI is', Generated text: ' 100% predictable. | by Aditya Bhargava |'
> Prompt: '希望这篇文章能', Generated text: ' 帮助您更好地了解我。 \n这篇文章将介绍我,一个名为'
> Prompt: '给六岁小朋友解释一下万有引', Generated text: '力的概念\n萬有引力是地球和其他物体之间的一种吸引力。它是由'

### speedup

it achieve a better speed up compared with `baichuan-inc/baichuan-7B` when tp=1

|checkpoint|backend|params|throughput|speedup|
|--|--|--|--|--|
|baichuan-inc/baichuan-7B|hf|hf_max_batch_size=4|136.01 tokens/s||
|baichuan-inc/baichuan-7B|vllm|tensor_parallel_size=1|1721.74 tokens/s|12.7|
|THUDM/chatglm3-6b|hf|hf_max_batch_size=4|204.21 tokens/s||
|THUDM/chatglm3-6b|vllm|tensor_parallel_size=1|4564.52 tokens/s|22.35|

### CODE REVIEW

**Comment by gameofdimension** [comment]
because the model architecture is the same as chatglm2-6b, so it also works for chatglm2-6b
> Prompt: 'Hello, my name is', Generated text: " Toots. I'm a member of the Hello, Toots community."
> Prompt: 'The president of the United States is', Generated text: '\xa0the head of the executive branch and has the power to make decisions and take'
> Prompt: 'The capital of France is', Generated text: ' In the capital of France, the population is approximately 12 million people.'
> Prompt: 'The future of AI is', Generated text: ' to become more human-like and to be able to perform tasks that currently require'
> Prompt: '希望这篇文章能', Generated text: '您的公司带来新的发展机遇。\n\n'
> Prompt: '给六岁小朋友解释一下万有引', Generated text: '力的概念。 首先，我们需要知道什么是引力。引力是物体之间的相互作用'

File: vllm/model_executor/models/chatglm.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+class MLP(nn.Module):
+
+    def __init__(self, config: ChatGLMConfig):
+        super().__init__()
+
+        def swiglu(x):
+            x = torch.chunk(x, 2, dim=-1)
+            return torch.nn.functional.silu(x[0]) * x[1]
+
+        self.activation_func = swiglu
```
Comment (cauyxy) [inline vllm/model_executor/models/chatglm.py:30]: Maybe use `SiluAndMul` instead
`from vllm.model_executor.layers.activation import SiluAndMul`

File: vllm/model_executor/models/chatglm.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        kv_cache: KVCache,
+        input_metadata: InputMetadata,
+        cache_event: Optional[torch.cuda.Event],
+    ) -> torch.Tensor:
+        hidden_states, _ = self.query_key_value(hidden_states)
+        query, key, value = hidden_states.split(
+            [
+                self.num_heads * self.head_dim,
```
Comment (cauyxy) [inline vllm/model_executor/models/chatglm.py:127]: These variables may be precomputed in __init__() as `q_size, kv_size, kv_size` like llama impl.
`q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)`

File: vllm/transformers_utils/configs/chatglm.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+from transformers import PretrainedConfig
```
Comment (cauyxy) [inline vllm/transformers_utils/configs/chatglm.py:1]: Add a copyright notice here?
`Copied from xxx` or `Adapted from xxx`

File: vllm/model_executor/models/chatglm.py (hunk 2 above)
Comment (cauyxy) [reply vllm/model_executor/models/chatglm.py:127]: `**hidden_states**, _ = self.query_key_value(hidden_states)`

maybe `hidden_states` is not a good variable name to hold the return value

**Comment by wengyuan722** [comment]
root@autodl-container-43e51183ae-e83acbf8:~/vllm-main/vllm/entrypoints/openai# python api_server.py --host 127.0.0.1 --port 6006 --model /root/autodl-tmp/chatglm3-6b --trust-remote-code --served-model-name chatglm3-6b --max-num-batched-tokens 8192 --max-model-len 8192
INFO 11-06 10:14:50 llm_engine.py:72] Initializing an LLM engine with config: model='/root/autodl-tmp/chatglm3-6b', tokenizer='/root/autodl-tmp/chatglm3-6b', tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float16, max_seq_len=8192, download_dir=None, load_format=auto, tensor_parallel_size=1, quantization=None, seed=0)
Traceback (most recent call last):
  File "api_server.py", line 623, in <module>
    engine = AsyncLLMEngine.from_engine_args(engine_args)
  File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/async_llm_engine.py", line 487, in from_engine_args
    engine = cls(engine_args.worker_use_ray,
  File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/async_llm_engine.py", line 270, in __init__
    self.engine = self._init_engine(*args, **kwargs)
  File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/async_llm_engine.py", line 306, in _init_engine
    return engine_class(*args, **kwargs)
  File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/llm_engine.py", line 96, in __init__
    self._verify_args()
  File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/llm_engine.py", line 187, in _verify_args
    self.model_config.verify_with_parallel_config(self.parallel_config)
  File "/root/miniconda3/lib/python3.8/site-packages/vllm/config.py", line 128, in verify_with_parallel_config
    total_num_hidden_layers = self.hf_config.num_hidden_layers
  File "/root/miniconda3/lib/python3.8/site-packages/transformers/configuration_utils.py", line 261, in __getattribute__
    return super().__getattribute__(key)
AttributeError: 'ChatGLMConfig' object has no attribute 'num_hidden_layers'

I use openai/api_server.py to start chatglm3-6b api,it return AttributeError: 'ChatGLMConfig' object has no attribute 'num_hidden_layers'

**Comment by gameofdimension** [comment]
> root@autodl-container-43e51183ae-e83acbf8:~/vllm-main/vllm/entrypoints/openai# python api_server.py --host 127.0.0.1 --port 6006 --model /root/autodl-tmp/chatglm3-6b --trust-remote-code --served-model-name chatglm3-6b --max-num-batched-tokens 8192 --max-model-len 8192 INFO 11-06 10:14:50 llm_engine.py:72] Initializing an LLM engine with config: model='/root/autodl-tmp/chatglm3-6b', tokenizer='/root/autodl-tmp/chatglm3-6b', tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float16, max_seq_len=8192, download_dir=None, load_format=auto, tensor_parallel_size=1, quantization=None, seed=0) Traceback (most recent call last): File "api_server.py", line 623, in engine = AsyncLLMEngine.from_engine_args(engine_args) File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/async_llm_engine.py", line 487, in from_engine_args engine = cls(engine_args.worker_use_ray, File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/async_llm_engine.py", line 270, in **init** self.engine = self._init_engine(*args, **kwargs) File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/async_llm_engine.py", line 306, in _init_engine return engine_class(*args, **kwargs) File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/llm_engine.py", line 96, in **init** self._verify_args() File "/root/miniconda3/lib/python3.8/site-packages/vllm/engine/llm_engine.py", line 187, in _verify_args self.model_config.verify_with_parallel_config(self.parallel_config) File "/root/miniconda3/lib/python3.8/site-packages/vllm/config.py", line 128, in verify_with_parallel_config total_num_hidden_layers = self.hf_config.num_hidden_layers File "/root/miniconda3/lib/python3.8/site-packages/transformers/configuration_utils.py", line 261, in **getattribute** return super().**getattribute**(key) AttributeError: 'ChatGLMConfig' object has no attribute 'num_hidden_layers'
> 
> I use openai/api_server.py to start chatglm3-6b api,it return AttributeError: 'ChatGLMConfig' object has no attribute 'num_hidden_layers'


maybe you are testing  against a wrong commit

File: vllm/model_executor/models/chatglm.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
+from typing import List, Optional, Tuple
```
Comment (cauyxy) [inline vllm/model_executor/models/chatglm.py:1]: add a copy right notice here？
and `Inference-only ChatGLM model compatible...` be consistent with other models.

**Review by cauyxy — COMMENTED** [review]
LGTM. Wait @zhuohan123  take a look at it.

File: vllm/config.py [hunk 5]
```diff
@@ -113,6 +113,12 @@ def _verify_quantization(self) -> None:
                 f"{supported_quantization}.")
         self.quantization = quantization
 
+    @staticmethod
+    def hf_config_get_num_layers(hf_config):
+        if getattr(hf_config, "model_type", None) == "chatglm":
+            return hf_config.num_layers
+        return hf_config.num_hidden_layers
```
Comment (zhuohan123) [inline vllm/config.py:120]: Please remove this function and add the following to `ChatGLMConfig`

```python
    attribute_map = {
        "num_hidden_layers": "num_layers",
    }
```

You can refer to falcon's config: https://github.com/vllm-project/vllm/blob/8516999495114926c9838c2d6e0feb580d4d983f/vllm/transformers_utils/configs/falcon.py#L25-L29

File: vllm/model_executor/layers/rotary_embedding.py [hunk 6]
```diff
@@ -40,13 +40,20 @@ def __init__(
         max_position_embeddings: int,
         base: int,
         is_neox_style: bool,
+        is_glm_style: bool = False,
```
Comment (zhuohan123) [inline vllm/model_executor/layers/rotary_embedding.py:43]: Can you change this to:
```suggestion
        style: str = "neox",
```
and set style to `"neox", "gptj", "glm"` accordingly?

File: vllm/model_executor/models/chatglm.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+from vllm.model_executor import InputMetadata
+from vllm.model_executor.layers.activation import SiluAndMul
+from vllm.model_executor.layers.attention import PagedAttentionWithRoPE
+from vllm.model_executor.layers.layernorm import RMSNorm
+from vllm.model_executor.layers.sampler import Sampler
+from vllm.model_executor.parallel_utils.layers import ColumnParallelLinear, RowParallelLinear, VocabParallelEmbedding
+from vllm.model_executor.parallel_utils.parallel_state import get_tensor_model_parallel_rank
+from vllm.model_executor.parallel_utils.parallel_state import get_tensor_model_parallel_world_size
+from vllm.model_executor.weight_utils import hf_model_weights_iterator, load_tensor_parallel_weights, \
+    load_padded_tensor_parallel_vocab
```
Comment (zhuohan123) [inline vllm/model_executor/models/chatglm.py:24]: Please use `()` for line continuation:
```suggestion
from vllm.model_executor.weight_utils import (
    hf_model_weights_iterator, load_tensor_parallel_weights, 
    load_padded_tensor_parallel_vocab)
```

File: vllm/model_executor/models/chatglm.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+
+    def __init__(self, config: ChatGLMConfig):
+        super().__init__()
+        self.rope_theta = 10000
+
+        tp_world_size = get_tensor_model_parallel_world_size()
+        self.total_num_heads, self.num_heads = compute_tp_num_heads(
+            config, tp_world_size)
+        self.head_dim = config.hidden_size // self.total_num_heads
+        self.total_num_kv_heads, self.num_kv_heads, num_kv_heads_replicas = \
```
Comment (zhuohan123) [inline vllm/model_executor/models/chatglm.py:92]: Please use `()` for line continuation.

File: vllm/model_executor/layers/attention.py [hunk 9]
```diff
@@ -311,6 +311,7 @@ def __init__(
         base: int = 10000,
         num_kv_heads: Optional[int] = None,
         is_neox_style: bool = True,
+        is_glm_style: bool = False,
```
Comment (zhuohan123) [inline vllm/model_executor/layers/attention.py:314]: Let's use a single style variable:
```suggestion
        style: str = "neox",
```

File: vllm/model_executor/models/chatglm.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+                weight2 = loaded_weight[base + shard_size * tp_rank:base +
+                                        shard_size * (tp_rank + 1)]
+                param.data[shard_size:].copy_(weight2)
+                continue
+
+            if 'query_key_value' in vname:
+                total_num_heads, num_heads = compute_tp_num_heads(
+                    self.config, tp_world_size)
+                head_dim = self.config.hidden_size // total_num_heads
+                total_num_kv_heads, num_kv_heads, num_kv_heads_replicas = \
```
Comment (zhuohan123) [inline vllm/model_executor/models/chatglm.py:301]: Please use `()` for line continuation.

File: vllm/transformers_utils/configs/chatglm.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        self.ffn_hidden_size = ffn_hidden_size
+        self.kv_channels = kv_channels
+        self.num_attention_heads = num_attention_heads
+        self.seq_length = seq_length
+        self.hidden_dropout = hidden_dropout
+        self.classifier_dropout = classifier_dropout
+        self.attention_dropout = attention_dropout
+        self.layernorm_epsilon = layernorm_epsilon
+        self.rmsnorm = rmsnorm
+        self.apply_residual_connection_post_layernorm = \
```
Comment (zhuohan123) [inline vllm/transformers_utils/configs/chatglm.py:50]: Please use `()` for line continuation.

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Thank you for your contribution! Left some style comments.

The PR runs correctly on my side on a single GPU. We should be able to merge this after the styles are fixed.

**Comment by zhuohan123** [comment]
Again, thank you for your contribution! We merged https://github.com/vllm-project/vllm/pull/1261 and the current main branch supports ChatGLM now. Let us know if the main branch does not look good and feel free to propose any changes!

**Comment by gameofdimension** [comment]
> Again, thank you for your contribution! We merged #1261 and the current main branch supports ChatGLM now. Let us know if the main branch does not look good and feel free to propose any changes!

pr 1261 might not support tp where world size is bigger than kv heads.

https://github.com/vllm-project/vllm/blob/1a2bbc930135cd3b94fbff2aafbdf5c568acc8bd/vllm/config.py#L171

**Pull request closed without merging**